File: spring-boot-3-demo-app/src/main/java/io/github/mfvanek/spring/boot3/test/controllers/TimeController.java
/*
 * Copyright (c) 2020-2025. Ivan Vakhrushev and others.
 * https://github.com/mfvanek/spring-boot-open-telemetry-demo
 *
 * Licensed under the Apache License 2.0
 */

package io.github.mfvanek.spring.boot3.test.controllers;

import io.github.mfvanek.spring.boot3.test.service.KafkaSendingService;
import io.github.mfvanek.spring.boot3.test.service.PublicApiService;
import io.micrometer.tracing.Span;
import io.micrometer.tracing.TraceContext;
import io.micrometer.tracing.Tracer;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Clock;
import java.time.LocalDateTime;
import java.util.Optional;

@Slf4j
@RestController
@RequiredArgsConstructor
public class TimeController {

    private final Tracer tracer;
    private final Clock clock;
    private final KafkaSendingService kafkaSendingService;
    private final PublicApiService publicApiService;

    // http://localhost:8080/current-time
    @SneakyThrows
    @GetMapping(path = "/current-time")
    public LocalDateTime getNow() {
        log.trace("tracer {}", tracer);
        final String traceId = Optional.ofNullable(tracer.currentSpan())
            .map(Span::context)
            .map(TraceContext::traceId)
            .orElse(null);
        log.info("Called method getNow. TraceId = {}", traceId);
        final LocalDateTime nowFromRemote = publicApiService.getZonedTime();
        final LocalDateTime now = nowFromRemote == null ? LocalDateTime.now(clock) : nowFromRemote;
        kafkaSendingService.sendNotification("Current time = " + now)
            .thenRun(() -> log.info("Awaiting acknowledgement from Kafka"))
            .get();
        kafkaSendingService.sendNotificationToOtherTopic("Current time = " + now)
            .thenRun(() -> log.info("Awaiting acknowledgement from Kafka with batch"))
            .get();
        return now;
    }
}

File: spring-boot-3-demo-app-kotlin/src/main/kotlin/io/github/mfvanek/spring/boot3/kotlin/test/TimeController.kt
/*
 * Copyright (c) 2020-2025. Ivan Vakhrushev and others.
 * https://github.com/mfvanek/spring-boot-open-telemetry-demo
 *
 * Licensed under the Apache License 2.0
 */
package io.github.mfvanek.spring.boot3.kotlin.test

import io.github.mfvanek.spring.boot3.kotlin.test.service.KafkaSendingService
import io.github.mfvanek.spring.boot3.kotlin.test.service.PublicApiService
import io.github.oshai.kotlinlogging.KotlinLogging
import io.micrometer.tracing.Tracer
import java.time.Clock
import java.time.LocalDateTime
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController

private val logger = KotlinLogging.logger {}

@RestController
class TimeController(
    private val tracer: Tracer,
    private val clock: Clock,
    private val kafkaSendingService: KafkaSendingService,
    private val publicApiService: PublicApiService
) {

    @GetMapping(path = ["/current-time"])
    fun getNow(): LocalDateTime {
        logger.trace { "tracer $tracer" }
        val traceId = tracer.currentSpan()?.context()?.traceId()
        logger.info { "Called method getNow. TraceId = $traceId" }
        val nowFromRemote = publicApiService.getZonedTime()
        val now = nowFromRemote ?: LocalDateTime.now(clock)
        kafkaSendingService.sendNotification("Current time = $now")
            .thenRun { logger.info{"Awaiting acknowledgement from Kafka"} }
            .get()
        return now
    }
}
